### ISSUE TITLE
Pedestrian Detection Model Training Problem

### ISSUE THREAD

**Issue by VigorPoppy** [eric612/MobileNet-YOLO#38]
Hi, I wanna create a pedestrian detection model(detect pedestrian, 1 class),  and I have prepared a dataset contains about 9000+ train images and 1000+ val images. when I use this dataset in both yolov3 or mobilenet-yolov3, I got avg_obj about 0.6 and avg_iou about 0.55, the loss value about 3 to 6, but each 1000 iters I test my model I got the detection_eval mAP was about 0.000* to 0.0*, a very samll number. This problem occured whenever I train my model for start or in the 5-8 epoch or in the end. So, could you ploese tell me what causes this problem probobly? I have checked my dataset and it's fine. Do you think resize the image in the aspect ratio level will cause this problem? thanks so much!

this is the train log
...
avg_noobj: 0.0010526 avg_obj: 0.400564 avg_iou: 0.762177 avg_cat: 0.987689 recall: 0.958333 recall75: 0.630952 count: 4
...
I1213 20:31:15.950788  6814 solver.cpp:548] class1: 0.000429772
I1213 20:31:15.950799  6814 solver.cpp:554]     Test net output #6: detection_eval_60 = 0.000429772

**Comment by eric612** [comment]
It do not like over fitting , can you provide train and test prototxt ? If have log file would be better.

**Comment by qiangqiu** [comment]
here is my train log, an online doc, I hope u can check it
[https://docs.qq.com/doc/DWHRhRHV4bFVwY3BV](url)

**Comment by eric612** [comment]
I can't see the number of "num_classes" both on loss and detection_eval. layer , can you also provide prototxt link ?

**Comment by qiangqiu** [comment]
this is the prototxt of train:
https://docs.qq.com/doc/DWGdEcWhTSEJhdU9s
and test：
https://docs.qq.com/doc/DWGZoQm1OWWdKU3R6

**Comment by qiangqiu** [comment]
I have also checked the num of class in train an test. and every output layers relative with the num_classes. Is the mobilenet_lite model I used too simple to learn from the data or too complex lead to the overfitting?

**Comment by eric612** [comment]
The prototxt was correct , have you been try the solver type to "RMSProp" ? I suggest use RMSProp and multistage param like [this](https://github.com/eric612/MobileNet-YOLO/blob/master/models/yolov3_coco/mobilenet_yolov3_lite_solver.prototxt) ,  although it would not effect your problem . 

Actually , I never try to train 1 class dataset , so I can not make sure what the problem. Thanks for your questions , I'll try to find the problems .

**Comment by qiangqiu** [comment]
I will have a try and thanks for your help.
For one-class model training, how about set class_scale to 0 in loss layer ? If I make any progress I’ll discuss with you in time!

**Comment by eric612** [comment]
Since the confidence = object score * class score , In my guess , 1 class detection would cause last term always large ? so eval. at conf = 0.1 may make something wrong . Or you can try draw the pictures down.

**Comment by qiangqiu** [comment]
hi, i have just finished my darknet-yolov3 training, there was problem below:
mAP was about 0.4 and I used example/yolo/detect.py to test the inference of my model, one of the result as show in this pic:
![aachen_000164_000019_leftimg8bit](https://user-images.githubusercontent.com/21115649/49989597-b07dd480-ffb5-11e8-8731-fb051f87d55c.jpg)
most of the result show just like the pic, and conf is about from 0.1 to 0.5

**Comment by eric612** [comment]
It looks like a open dataset , would you mind give me the download link ?

**Comment by qiangqiu** [comment]
yeah. it's called CityPerson(annotations: https://bitbucket.org/shanshanzhang/citypersons), a re-annotated data from CityScapes, Its contain pedestrian riders and sitting person ... my whole dataset include KITTI(person) ETH and CityPerson, I remove all the bbox which  less than 20 pixels(width or h). I modified all the categories to pedestrians(riders sitting_person..) because KITTI only have pedestrian annotation(including riders). each image in CityPerson was 1024*2048, so I cut this image into two parts I hope that my model will performance well.

**Comment by eric612** [comment]
I have some problems to download the CityScapes dataset , but I try convert my dataset to one class first. It seems work fine. So I guess maybe it's not 1 class problem. Did you resize images when create lmdb, the ssd library only support original image size and anno. so on . 

My dataset training log
https://drive.google.com/open?id=1CCxYjLFXXzCIWBZJexhbG-rlwKz-djHK

![image](https://user-images.githubusercontent.com/15793472/50061572-a8ec4480-01dc-11e9-85c8-f8e9c986c67f.png)

**Comment by qiangqiu** [comment]
I set width and height both equal to 0 and there was different size of images in my dataset. And  I thought 0 means it maybe set image size as default(i.e. original image size)

**Comment by eric612** [comment]
Yes , you right . I have no idea , now .

**Comment by qiangqiu** [comment]
I thought that the problom may caused by my dataset. There are so many ccrowded objects in image. The train is not stable as shoewd below(iters12000):
#Iters    Seconds      TestAccuracy
1000   2490.765750   0.38002
2000   4538.115014   0.392588
3000   6580.004820   0.401048
4000   8641.068245   0.409091
5000   10691.136725  0.395815
6000   12737.417135  0.412718
7000   14788.022421  0.405816
8000   16830.844980  0.396541
9000   18869.035031  0.398855
10000  20914.411404  0.413621
11000  22954.299325  0.396963
12000  24999.145399  0.000453412
13000  27057.200902  0.139452
14000  29104.155054  0.102807
15000  31139.768798  0.249749
16000  33196.923409  0.289762
17000  35248.799999  0.348935
18000  37287.980031  0.292469
19000  39330.211928  0.365838
20000  41367.340803  0.202436
21000  43411.914032  0.286503
22000  45449.627884  0.298282
23000  47496.596385  0.310212
24000  49553.129934  0.330583
25000  51593.303792  0.339155

**Comment by eric612** [comment]
The yolo-lite does not provide enough bounding box , you may need try three scales detector , or fixed input resolution at 608.

**Comment by qiangqiu** [comment]
yeah I found that yolo performance better than yolo-lite(test mAP about 0.61 vs. 0.27), I thought that I should add some conv. layers and one more detect layer. Or could I modifiy the parameter num in the detect layer from 3 to 5, add two bbox in each layer?

**Comment by eric612** [comment]
* Maybe you can train yolo-lite with pre-trained weights with yolo (3 scales) , I think the mAP can be arrive to 5x mAP. 
* Or as you say , try to add more bounding box detector 
Note : here has two ways , one way is modify num 3 to 5 , or add yolo layer at scale 1/16 , which will not increase much computation

**Comment by qiangqiu** [comment]
The 3scale mode based on yolov3_mobilenet_lite_coco（http://ethereon.github.io/netscope/#/gist/110f5f5a2edad80c0c9074c7a532347b） is under training but it only evaluated a mAP 0.2. This is my modified model（http://ethereon.github.io/netscope/#/gist/5b6deecb3ee07f94139297baac7fe134）. Could you give me some advice?

**Comment by eric612** [comment]
Sorry , I miss the point , did you mean you want add third yolo layer on first model ?
And the mAP of second model is the last you reply said 0.61?

**Comment by qiangqiu** [comment]
yes,sure. the mAP of [http://ethereon.github.io/netscope/#/gist/5b6deecb3ee07f94139297baac7fe134%EF%BC%89](second model) is about 0.2, which means there is no effect on yolo-lite with 3 scale(1/8 1/16 1/32, which 1/8 is the new scale in yolo-lite I added)

**Comment by eric612** [comment]
Okay , can you give me the 0.61 mAP models ? start from here I think will be faster

**Comment by qiangqiu** [comment]
this is [train.prototxt](https://drive.google.com/open?id=1xpi2oiZJN9v_IJ4O0mLAQCdiY_cNWBCy)、[test.prototxt](https://drive.google.com/open?id=1dV3HE-g3TFTuqiWXzmaxCeyKbbomHXw2) and [solver.prototxt](https://drive.google.com/open?id=1OVWEBYbUF5MghjaZaY0k0wC8Zc4WviZh)

**Comment by eric612** [comment]
Okay ... Maybe you need some steps to find the root cost
* use [pre-trained weights](https://drive.google.com/file/d/0B3gersZ2cHIxZi13UWF0OXBsZzA/view) or others 
* try different data augmentation  (expand param(different scales) , batch sampler(instead of yolo jitter))
* Or if you can you can give me the lmdb (or particle lmdb), it is fastest .
* I also recommend you can try to train mobilenet-ssd at resolution 300 and 512 , it can help to find the problems

**Comment by eric612** [comment]
I receive the prototxt you send , that the batch size = 8x2 = 16 , can you set iter to 32 and retrain ?

**Comment by qiangqiu** [comment]
This is the [train lmdb](https://drive.google.com/open?id=1sv56ZnaWvKe6aUKcxE3DM7RazoM3gBmf) and [test lmdb](https://drive.google.com/open?id=1fdEnLON8sIKwBO0vFgZWQ5XoIPuw35FK)

**Comment by qiangqiu** [comment]
And there is another question. I wanna check my test loss so that I add yololoss layer in test.prototxt. But an error occured "check failed: result[j]->width() ==5 (1 vs 5)". I have tried to add silence layer and other methods, it didnot work well. I checked the issues in SSD(luwei) and I found that many people had same question. So do you have some ideas about how to modify the test.prototxt?

**Comment by eric612** [comment]
How about set test lmdb path in training prototxt (resume in specify iter solverstate)? It also can test convergence problem in test lmdb

**Comment by qiangqiu** [comment]
Terrific idea! Thakns!

**Comment by eric612** [comment]
I use the lmdb that you provide , but I meet below problem  

```
E1222 23:56:23.710422  6696 data_transformer.cpp:581] Only support encoded datum now
F1222 23:56:23.720664  6696 data_transformer.cpp:200] Check failed: height <= datum_height (576 vs. 480)
```

**Comment by VigorPoppy** [comment]
Thanks for your help and suggestions！
